### PR TITLE
Rework build logic

### PR DESCRIPTION
**Pull request by whyoleg**
This PR is rather big, but it will be hard to split it further...

### Modifications:

* add `build-parameters` module for typesafe parameters access in gradle scripts
  * not everything was migrated yet, but will be migrated some day later
* add `kotlin-version-catalog` module for easy overriding of kotlin version
  * this is a requirement for upcoming PR which will introduce testing building over different kotlin version and K2 compiler. This will help to catch bugs earlier and report them
* replace buildSrc with `build-logic` module
* use convention plugins instead of custom DSL for targets configuration
  * current custom DSL confuses a lot. Using convention plugins with self-descriptive names and default gradle functions to configure source sets and dependencies is easier to maintain
* use .configureEach instead of .all
* use new kotlin `targetHierarchy` API to set up HMPP source sets
* apply `kotlinx-atomicfu` to only modules where it's needed (core, test, transport-tests)
* run tests also on JDK 20 (latest)

### CODE REVIEW

**Comment by whyoleg** [comment]
@yschimke hey, could you do some small review/walkthrough? :)

File: gradle/plugins/build-logic/settings.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+dependencyResolutionManagement {
+    repositories {
+        gradlePluginPortal()
+        mavenCentral()
+    }
+
+    versionCatalogs {
+        create("libs") {
```
Comment (yschimke) [inline gradle/plugins/build-logic/settings.gradle.kts:32]: Why is this needed?

File: gradle/plugins/build-logic/src/main/kotlin/rsocket.multiplatform.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        compilations.configureEach {
+            compilerOptions.configure {
+                freeCompilerArgs.add("-Xrender-internal-diagnostic-names")
+            }
+        }
+    }
+
+    sourceSets.configureEach {
+        languageSettings {
+            progressiveMode = true
```
Comment (yschimke) [inline gradle/plugins/build-logic/src/main/kotlin/rsocket.multiplatform.gradle.kts:40]: What is progressive mode?

**Review by yschimke — APPROVED** [review]
Above my head, but seems legit

File: gradle/plugins/build-logic/src/main/kotlin/rsocket.multiplatform.gradle.kts (hunk 2 above)
Comment (whyoleg) [reply gradle/plugins/build-logic/src/main/kotlin/rsocket.multiplatform.gradle.kts:40]: https://kotlinlang.org/docs/compiler-reference.html#progressive
Mostly to see compiler errors earlier, when they are introduced as warnings

File: gradle/plugins/build-logic/settings.gradle.kts (hunk 1 above)
Comment (whyoleg) [reply gradle/plugins/build-logic/settings.gradle.kts:32]: because we have single version catalog file in `gradle/libs.versions.toml`, and in included builds, we need to specify it explicitly, because by default it will search in `included-build-project-root/gradle/libs.versions.toml`.
